src/journal/journald-context.c

```diff
@@ -268,6 +273,46 @@ static int client_context_read_label(
         return 0;
 }
 
+static int client_context_read_log_filter_patterns(ClientContext *c, const char *cgroup) {
+        const char *pattern;
+        _cleanup_free_ char *xattr = NULL;
+        _cleanup_set_free_ Set *allow_list = NULL, *deny_list = NULL;
+        Set **current_list = &allow_list;
+        int r;
+
+        r = cg_get_xattr_malloc(SYSTEMD_CGROUP_CONTROLLER, cgroup, "user.journald_log_filter_patterns", &xattr);
+        if (r < 0 && r != -ENODATA)
+                return log_debug_errno(r, "Failed to get user.journald_log_filter_patterns xattr for %s: %m", cgroup);
+
+        NULSTR_FOREACH(pattern, xattr) {
+                pcre2_code *compiled_pattern = NULL;
+
+                if (pattern[0] == (char)0xff) {
+                        current_list = &deny_list;
+                        continue;
+                }
+
+                r = pattern_compile_and_log(pattern, 0, &compiled_pattern);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to compile log filtering pattern '%s' for %s: %m",
+                                               pattern, cgroup);
+
+                r = set_ensure_put(current_list, NULL, compiled_pattern);
+                if (r < 0) {
+                        compiled_pattern = pattern_free(compiled_pattern);
+                        return log_debug_errno(r, "Failed to insert regex into set for %s: %m", cgroup);
+                }
+        }
+
+        c->log_filter_allowed_patterns = set_free_with_destructor(c->log_filter_allowed_patterns, pattern_free);
+        c->log_filter_denied_patterns = set_free_with_destructor(c->log_filter_denied_patterns, pattern_free);
```

Comment:
actually, for new code we prefer if you'd just use set_free(), and specify a destructor for the data `via struct hash_ops` at allocation time.